keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs/README.md

```diff
@@ -0,0 +1,1212 @@
+Job is marked Failed and its execution is terminated.
+
+We also propose to extend the PodFailurePolicy with a new action, called
+`FailIndex` to allow an index to fail fast before reaching the backoff limit
+per index.
+
+### Job API
+
+```golang
+
+// PodFailurePolicyAction specifies how a Pod failure is handled.
+// +enum
+type PodFailurePolicyAction string
+
+const (
+  // This is an action which might be taken on a pod failure - mark the
+  // Job's index as failed to avoid restarts within this index. This action
+  // can only be used when backoffLimitPerIndex=true.
+  PodFailurePolicyActionFailIndex PodFailurePolicyAction = "FailIndex"
+  ...
+)
+...
+
+// JobRunPolicy describes custom policy for running a Job
+type JobRunPolicy struct {
+
+	// Specifies if backoffLimit is used to limit the number of retries within an
+  // index before marking this index as failed. When enabled the number of
+  // failures per index is kept in the pod's
+  // batch.kubernetes.io/job-index-failure-count annotation. It can only
+  // be set to true when Job's completionMode=Indexed, and the Pod's restart
+  // policy is Never.
+  // The value is immutable.
+  // Defaults to false
+  // +optional
+  BackoffLimitPerIndex *bool
+
+  // Specifies the maximal number of failed indexes before marking the Job as
+  // failed, when backoffLimitPerIndex=true. Once the number of failed
+  // indexes exceeds this number the entire Job is marked as Failed and its
+  // execution is terminated. When left as nil the job continues execution of
+  // all of its indexes and is marked with the `Complete` Job condition.
+  // It can only be specified when backoffLimitPerIndex=true.
+  // The value is might be in range from 0 to 10^5.
+  // +optional
+  MaxFailedIndexes *int32
+}
+
+// JobSpec describes how the job execution will look like.
+type JobSpec struct {
+  ...
+
+  // Specifies the policy for running a Job. In particular, it allows to enable
+  // the backoff limit per index.
+  // +optional
+  RunPolicy *JobRunPolicy
+  ...
+}
+
+type JobStatus struct {
+  ...
+
+  // FailedIndexes holds the failed indexes when
+  // `.spec.runPolicy.backoffLimitPerIndex=true`.
+  // The indexes are represented in the text format analogous as for the
+  // `completedIndexes` field, ie. they are kept as decimal integers
+  // separated by commas. The numbers are listed in increasing order. Three or
+  // more consecutive numbers are compressed and represented by the first and
+  // last element of the series, separated by a hyphen.
+  // For example, if the failed indexes are 1, 3, 4, 5 and 7, they are
+  // represented as "1,3-5,7".
+  // +optional
+  FailedIndexes *string
+}
+```
+
+### Tracking the number of failures per index
+
+In order to determine if the backoff limit per index is exceeded we keep
+track of the number of failures per index. For this purpose we use the Pod
+annotation, `batch.kubernetes.io/job-index-failure-count`, which holds the value
```

Comment:
If https://github.com/kubernetes/enhancements/issues/3322 supported, will it reduce the maintenance cost of Job api.

Reply:
I think we will be better of indeed if the container restarts are managed by Kubelet (when maxRestartTimes is used). Then, the regular backoffLimit can play a role of maxFailedIndexes.

Reply:
If we do have the usecases, I think we can redesign it from the scratch(maxRestartTimes respects restartPolicy=Never). But it needs to be discussed further. So the general idea is:
If we support maxRestartTimes in pod.Spec, which respects all restartPolicy, then we can sink the capacities of backOffLimitPerIndex to the pod level, the main difference here is Job controller will recreate the pods when restartPolicy=Never, currently, pod is searching for a new restart policy as recreate itself rather than restart, if it's implemented, than seems all the gaps are gone?

Reply:
> IIUC that linked KEP is for OnFailure pods, which are explicitly excluded from this proposal, at least initially, see non goals. When the linked proposal reaches GA we can consider expanding the scope of this feature.

Yes, if the linked KEP (`maxRestartTimes` is supported at the pod level), the we could re-asses the scope of this KEP if maxRestartTimes is available.

Reply:
> So the general idea is:
If we support maxRestartTimes in pod.Spec, which respects all restartPolicy, then we can sink the capacities of backOffLimitPerIndex to the pod level

I think that when `restartPolicy: Never`, then it is Job controller that needs to handle the backoffLimitPerIndex, as the pod is deleted / recreated. It is only when `restartPolicy: OnFailure`, when we can delegate controlling the number of restarts to kubelet.

Reply:
Note that kubelet can only do so much. For example, it cannot keep count of the number of restarts after a pod preemption, because there would be a new pod.